### PR TITLE
Adds DBT compare subcommand for comparing local models with database catalog

### PR DESCRIPTION
**Pull request by mikekaminsky**
Addresses https://github.com/fishtown-analytics/dbt/issues/1135

Here's an example:

``` 
$ dbt compare

Running with dbt=0.13.0-a1
Comparing local models to the database catalog. Checking schemas:
- dev_example
- dev_downstream
Warning: The following relations do not match any models found in this project:
TABLE "dev_example"."b"
VIEW "dev_downstream"."d"
```

### CODE REVIEW

File: core/dbt/adapters/base/impl.py [hunk 1]
```diff
@@ -687,6 +687,18 @@ def _catalog_filter_table(cls, table, manifest):
         """
         return table.where(_catalog_filter_schemas(manifest))
 
+    def get_unfiltered_catalog(self, manifest):
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/adapters/base/impl.py]: For some reason I had thought this was filtering at the relation level, not the schema level. Good catch!

File: core/dbt/main.py [hunk 2]
```diff
@@ -489,7 +497,7 @@ def parse_args(args):
         help='Do not run "dbt compile" as part of docs generation'
     )
 
-    for sub in [run_sub, compile_sub, generate_sub]:
+    for sub in [run_sub, compile_sub, generate_sub, compare_sub]:
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/main.py]: 👍 for some reason this had fixed some issue that I was experiencing when I was first getting started and I forgot to come back and remove this :)

File: core/dbt/task/compare.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    def run(self):
+
+        logger.info('dbt compare: Comparing local codebase to catalog in data warehouse.')
+
+        manifest = self._get_manifest()
+
+        # Look up all of the relations dbt knows about
+        used_schemas = manifest.get_used_schemas()
+        used_relations = []
+        for node in manifest.nodes.items():
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/task/compare.py]: I updated to

[the remark this answers is not in the document]
Comment (drewbanin) [reply core/dbt/task/compare.py]: Hmmm, I brought up ephemeral models because it's pretty typical to reconfigure eg. a view model to become an ephemeral model. When that happens, there's an orphaned `view` which will never be updated by dbt. I think that `dbt compare` could be really useful in determining when that happens!

So, while the ephemeral model is indeed _in_ the manifest, it won't actually be rendered out into the db. I think that makes it fall into your first point:
> it looks for things in the database that are not present in the manifest

We might as well throw away any ephemeral models (as we do with disabled models) when computing the difference between the catalog and the manifest i think.

This is a huge part of why i was so hesitant to make dbt actually drop the orphaned relations in the database. You can imagine code like:

```
{% if target.name == 'prod' %}

-- really complicated and involved incremental logic here
-- for a massive table that you don't want to build in dev

{% else %}

{{ config(materialized='ephemeral') }}

select * from prod.big_table

{% endif %}
```

If you're not careful with your targets, then dbt could errantly tell you to drop this model!

**Comment by drewbanin** [comment]
**What's the appropriate way to print / log for this feature?**

I think `logger.info` is appropriate. I could also imagine a flag to print this data in some sort of structured format that can be more readily consumed by some sort of script. That can also use `logger.info`, but with eg. JSON output.

**How to test it?**

This might be tricky. I'd make a new integration test that results in one or two relations in the database with no corresponding models in the manifest. Would be good to cover the enabled/disabled, ephemeral cases above, as well as things like custom schemas / aliases. There's really a lot of surface area for bugs here (due to the nature of how dbt builds relations,), so the more we can test, the better!

This test might be a good starting point:
https://github.com/fishtown-analytics/dbt/blob/dev/stephen-girard/test/integration/007_graph_selection_tests/test_graph_selection.py

**Should the comparison be bi-directional and also look for things in the manifest that aren't in the database?**

I don't think this part is necessary, but would be open to hearing use cases if you have them!

------

Running these tests on all adapters is going to be important I think. On Snowflake, relations are uppercased by default. I think that will cause the logic in this PR to report _all_ relations as missing from the manifest, even though they're just cased incorrectly. Don't worry about that part for now, but it's something to be mindful of for sure

File: core/dbt/task/compare.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/task/compare.py]: @drewbanin got it! This makes sense. Will implement.

**Comment by mikekaminsky** [comment]
@drewbanin cool, I'll work on the tests today -- do you have any preferences on how the output is formatted?

File: core/dbt/task/compare.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/task/compare.py]: Okay, I believe this is done now. On to the tests...

**Comment by mikekaminsky** [comment]
@drewbanin I added an integration test that works with postgres :)

I _believe_ I've also added the code to run the tests for redshift / bigquery / snowflake but I haven't ... tested the tests yet. Can you give it a whirl and see what happens (assuming it doesn't look too crazy to you)?

**Comment by drewbanin** [comment]
Hey @mikekaminsky - if you check out the Appveyor logs, you should see an error like this:
```
======================================================================
FAIL: test__bigquery__compare (test_compare.TestConfigs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\dbt\test\integration\base.py", line 667, in func
    return wrapped(self, *args, **kwargs)
  File "C:\projects\dbt\test\integration\040_compare_test\test_compare.py", line 68, in test__bigquery__compare
    self.compare_switch_to_ephemeral()
  File "C:\projects\dbt\test\integration\040_compare_test\test_compare.py", line 33, in compare_switch_to_ephemeral
    self.run_dbt(['run'])
  File "C:\projects\dbt\test\integration\base.py", line 360, in run_dbt
    "dbt exit state did not match expected")
AssertionError: False != True : dbt exit state did not match expected
-------------------- >> begin captured logging << --------------------
```

via: https://ci.appveyor.com/project/DrewBanin/dbt/builds/21451678

**Comment by mikekaminsky** [comment]
@drewbanin I've been doing some trouble-shooting today -- I thought I had finally gotten it to a good place but the appveyor run timed out (without erroring). I don't see anything particularly incriminating in the logs and am not sure how to trouble-shoot.

**Comment by drewbanin** [comment]
@mikekaminsky dang, yeah, Appveyor sometimes times out if the Snowflake build takes too long. Is this generally in a place where things are working for pg/redshift? I'm happy to give it a spin on a Snowflake/BQ instance and update here with some feedback if so

**Comment by mikekaminsky** [comment]
@drewbanin yep it seems to be working for postgres / redshift without issues. Let's see how this appveyor run goes but I may need some help wrapping the PR up since I don't have a bigquery / snowflake instance to test it on. 

Probably could use a bit more test coverage (for seeds and disabled models) but wanted to get it working with the basic tests first (and potentially have you do another round of review) before starting on that.

**Comment by mikekaminsky** [comment]
@drewbanin failed with

File: core/dbt/task/compare.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        # Look up all of the relations in the DB
+        adapter = get_adapter(self.config)
+        results = adapter.get_catalog(manifest)
+
+        results = [dict(zip(results.column_names, row)) for row in results]
+
+        existing_relations = {}
+        for x in results:
+            name = "%s.%s" % (x["table_schema"], x["table_name"])
+            existing_relations[name] = x["table_type"]
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/task/compare.py]: I was using the table type for the logging / output, but am not committed to keeping it :)

File: core/dbt/task/compare.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+        logger.info(
+            "dbt compare: Comparing local codebase to catalog in data warehouse."
+        )
+
+        manifest = self._get_manifest()
+
+        # Look up all of the relations dbt knows about
+        used_schemas = manifest.get_used_schemas()
+        used_relations = []
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/task/compare.py]: Good suggestion!

File: core/dbt/task/compare.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            name = "%s.%s" % (x["table_schema"], x["table_name"])
+            existing_relations[name] = x["table_type"]
+
+        problems = {}
+        for k, v in existing_relations.items():
+            if k.lower() not in used_relations:
+                problems[k] = v
+
+        if len(problems) == 0:
+            logger.info("All clear! The catalog matches the manifest.")
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/task/compare.py]: updated to

File: core/dbt/task/compare.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+        all_projects = compiler.get_all_projects()
+
+        manifest = dbt.loader.GraphLoader.load_all(self.config, all_projects)
+        return manifest
+
+    def run(self):
+
+        logger.info(
+            "dbt compare: Comparing local codebase to catalog in data warehouse."
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/task/compare.py]: Good idea!

File: core/dbt/task/compare.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply core/dbt/task/compare.py]: Of course happy to workshop these messages if you have ideas for clarity :)

File: test/integration/040_compare_test/test_compare.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    @staticmethod
+    def dir(path):
+        return "test/integration/040_compare_test/" + path.lstrip("/")
+
+    @property
+    def models(self):
+        return self.dir("models")
+
+    def create_test_model(self, materialization):
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply test/integration/040_compare_test/test_compare.py]: I left this as is for now. Happy to make the change but didn't want to go through the pain if we're not sure the right way to do it :)

**Comment by mikekaminsky** [comment]
@drewbanin omg the tests pass

**Comment by drewbanin** [comment]
@mikekaminsky nice! This works super well in most cases as is _really_ close! I'm going to drop a couple more comments in the code. After one more round of updates, I can kick off the full test suite and we can get this thing merged :)

File: core/dbt/task/compare.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        results = [dict(zip(results.column_names, row)) for row in results]
+
+        database_relations = set()
+        for row in results:
+            rel = (row["table_schema"].lower(), row["table_name"].lower())
+            database_relations.add(rel)
+
+        checked_schemas = set([x[0] for x in database_relations])
+
+        logger.info("-" * 40)
```
Comment (drewbanin) [inline core/dbt/task/compare.py]: I think this is a little heavy on the dashes. Example output looks like:

```
dbt compare: Comparing local codebase to catalog in data warehouse.
----------------------------------------
dbt compare reviewed the following schemas:
dbt_dbanin
----------------------------------------
----------------------------------------
Warning: We found some discrepancies between the database catalog and the dbt models:
This may indicate that you have outdated relations in your warehouse that can be removed.
----------------------------------------
DIFFERENCE: dbt_dbanin.events
DIFFERENCE: dbt_dbanin.fullevents_showsheet
DIFFERENCE: dbt_dbanin.fullevents_login
DIFFERENCE: dbt_dbanin.fullevents_view
```

We don't have any hard and fast rules about dbt's command line UX, but I do like to keep things mostly consistent when possible. Can you reswizzle this to look something like:

```
Comparing local models to the database catalog. Checking schemas:
- dbt_dbanin
- dbt_dbanin_abc
- dbt_dbanin_def

Warning: The following relations do not match any models found in this project:
table "dbt_dbanin"."events"
view "dbt_dbanin"."fullevents_show"
table "dbt_dbanin"."fullevents_login"
table "dbt_dbanin"."fullevents_view"
```

There are a couple of subtle changes here that i think are important:
1) Use a formal 3rd person tense rather than we/you. I'm not actually opposed to the style you used here, but it's not something we do elsewhere in dbt AFAIK, and I think it's important to be consistent
2) Remove the `DIFFERENCE` from each problem line. This will make it easy to copy/paste these things into an editor where you can drop them.
3) Add the relation type back in. Sorry for leading you astray earlier -- i didn't realize this is where you were using the relation type, and I think it's a super good idea. I like that you can prepend all of these with `drop` and get some runnable queries (if you so choose)
4) Use quoted relation names (more on that below)

If you want to get cute, I think you can use `dbt.ui.printer.green()` and `dbt.ui.printer.yellow()` for the status messages :)

File: core/dbt/task/compare.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            is_refable = node["resource_type"] in NodeType.refable()
+            is_enabled = check_is_enabled(node)
+            is_ephemeral = node["config"]["materialized"] == "ephemeral"
+            if is_refable and is_enabled and not is_ephemeral:
+                rel = (node["schema"].lower(), node["alias"].lower())
+                model_relations.add(rel)
+
+        # Look up all of the relations in the DB
+        adapter = get_adapter(self.config)
+        results = adapter.get_catalog(manifest)
```
Comment (drewbanin) [inline core/dbt/task/compare.py]: I took a deeper look at this and don't think we actually want to use `get_catalog` here. That method returns the full catalog, which includes all of the relations in the specified schemas, PLUS all of the columns in those relations. Instead, you should be able to just use `list_relations`, eg:

```
relations = []
for schema in manifest.get_used_schemas():
  relations.extend(adapter.list_relations(schema))
```

This will make `results` a list of [Relation](https://docs.getdbt.com/reference#relation) objects. Relations are great because they know how to quote themselves (see below):

```
ipdb> my_relation.schema
'demo_schema'
ipdb> my_relation.identifier
'my_table'
ipdb> my_relation
<BaseRelation "demo_schema"."my_table">
ipdb> str(my_relation)
'"demo_schema"."my_table"'
```

I'd do two things with these relations:
1) convert them into a tuple (as you already do below)
2) add an entry to a dictionary which maps the tuple onto the original relation

Eg:
```
database_relations = set()
database_relations_map = dict()
for relation in relations:
    relation_id = (relation.schema.lower(), relation.identifier.lower())
    database_relations_map[relation_id] = relation
    database_relations.add(relation_id)
```

If you do this, then you can do a set difference below to calculate the extraneous db relations, then _look up the original relation object_. This will help with effective quoting, again: see below

File: core/dbt/task/compare.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+            logger.info(
+                "Warning: We found some discrepancies between the database catalog and the dbt models:"
+            )
+            logger.info(
+                "This may indicate that you have outdated relations in your warehouse that can be removed."
+            )
+            logger.info("-" * 40)
+
+        for relation in problems:
+            logger.info("DIFFERENCE: %s.%s" % (relation[0], relation[1]))
```
Comment (drewbanin) [inline core/dbt/task/compare.py]: Quoting is a huge PITA on Snowflake in particular, and this implementation is close, but could be made a little more accurate. On Snowflake, identifiers are automatically uppercased if left unquoted. If you quote an identifier (like a schema/table/column name), then Snowflake will preserve the casing of your identifier.

More info on this phenomenon [here](https://docs.getdbt.com/docs/configuring-quoting#section-snowflake). I will leave aside my judgement on this particular feature of Snowflake, but will note that we have to account for this pretty much everywhere we render out a relation name in dbt. As such, I think it makes sense to use the quoted version of the relation from above.

If you do build `database_relations_map`, then you can do:
```
for relation_id in problems:
    relation = database_relations_map[relation_id]
    logger.info("{} {}".format(relation.type, relation))
```

File: core/dbt/task/compare.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        logger.info("-" * 40)
+
+        return problems
```
Comment (drewbanin) [inline core/dbt/task/compare.py]: Can you make this return a list of Relations? That will be useful if someone wants to invoke this task via Python and then do something with the results. 

Also: if you implement the [interpret_results](https://github.com/fishtown-analytics/dbt/blob/dev/stephen-girard/core/dbt/task/base_task.py#L12) method, you can make this exit with a nonzero exit code if problems exist. That would look like:

```
def interpret_results(self, problems):
    return len(problems) == 0
```

Basically, the invocation was successful if there were no problems

**Comment by mikekaminsky** [comment]
@drewbanin love all of your suggestions. I took all of them pretty much verbatim (though I upper-cased the relation in the output). 

```
Running with dbt=0.13.0-a1
Comparing local models to the database catalog. Checking schemas:
- dev_example
- dev_downstream
Warning: The following relations do not match any models found in this project:
TABLE "dev_example"."b"
VIEW "dev_downstream"."d"
```

**Comment by mikekaminsky** [comment]
But now that I think about it, I think there might be a bug waiting to happen here. This is valid (at least in postgresql):

```
CREATE SCHEMA "SchemaName";
CREATE SCHEMA "schemaname";
```

So our `relation_id` isn't right because it lower-cases the schema and relation name. What do you think? Does the `str(relation)` trick handle this sort of capitalization appropriately so that we can use _that_ as the `relation_id`?

**Comment by drewbanin** [comment]
dbt actually doesn't permit you to have two relations with "ambiguous" names in a dbt-controlled schema. We say that two names are ambiguous if their schema.table differs only in capitalization. 

You can test this out by creating two models in your own project: `TEST_IT.sql` and `test_it.sql`. You should see an error like:
```
Compilation Error in model test_it (models/test_it.sql)
  When searching for a relation, dbt found an approximate match. Instead of guessing
  which relation to use, dbt will move on. Please delete "demo_schema"."TEST_IT", or rename it to be less ambiguous.
  Searched for: "demo_schema"."test_it"
  Found: "demo_schema"."TEST_IT"

  > in macro materialization_view_default (macros/materializations/view/view.sql)
  > called by model test_it (models/test_it.sql)
```

All of this exists precisely to work around Snowflake's unquoted capitalization wonkiness 🙃. So, we can safely assume that there won't be any collisions if we lower-case everything.

I didn't think that we could use Relations as dict keys, but it actually looks like that's permitted! For something to be a dict key, it needs to be "hashable", and it does look like the Relation implements a [hash](https://github.com/fishtown-analytics/dbt/blob/dev/stephen-girard/core/dbt/adapters/base/relation.py#L216) method.

This is one where I'd defer to @beckjake. Jake, mind taking a look at this PR?

**Comment by beckjake** [comment]
Sorry, where is this PR using relations as dict keys? I only see `(schema.lower(), identifier.lower())` which seems correct (in the absence of database awareness).

The only issue with using relations as dict keys is you have to be very careful about the disconnect between hash collisions that have different actual values. `key in dict` can be True while `any(key == k for k in dict)` is False, which violates a lot of peoples' expectations!

Plus the same relation with different quoting will get hashed differently, since the hash function preserves the quotes and casing.

File: core/dbt/task/compare.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        database_relations = set()
+        database_relations_map = dict()
+        for relation in db_relations:
+            relation_id = (relation.schema.lower(), relation.identifier.lower())
+            database_relations_map[relation_id] = relation
+            database_relations.add(relation_id)
+
+        logger.info("Comparing local models to the database catalog. Checking schemas:")
+        for schema_name in checked_schemas:
+            logger.info("- %s" % schema_name)
```
Comment (drewbanin) [inline core/dbt/task/compare.py]:
```suggestion
            logger.info("- {}".format(schema_name))
```

**Comment by mikekaminsky** [comment]
hey @beckjake the ask here wasn't quite clear enough, I think :)

I'll try to re-state the question in general terms: currently, in order to 'compare' the database catalog with the dbt manifest, we create lists of schema-qualified relations from both sources. We lower-case those schema-qualified relation names in all cases.

We're speculating as to whether this could cause issues in cases where people have case-sensitive schema or relation names. Drew's thought is that since this isn't compatible with DBT we shouldn't worry about it, and we both speculated (though didn't implement) about using some other type of identifier for identifying relations.

What do you think? Is this good as-is (with lower-cased relation names) or do we need to do something more sophisticated?

**Comment by beckjake** [comment]
Force-lowercasing is fine, I think, as long as it's for stuff that comes from dbt - as drew mentioned, dbt makes sure you don't have any duplicates during a run. There is a false-positive issue where if an uppercase model existed and dbt would generate a lowercase version (due to quoting on snowflake or something), we'd compare them as equal. But that sounds ok to me, lowercasing is what we did on the cache as well and it has made life a lot easier. Case-sensitively comparing relation names across multiple datastores is unbelievably difficult!

**Comment by mikekaminsky** [comment]
Cool! I say we release as-is and see if people run into any issues.

**Comment by drewbanin** [comment]
@mikekaminsky just kicked off the full test suite. I think you're going to see some PEP8 errors. Can you address those if/when they pop up?

**Comment by mikekaminsky** [comment]
Ah yes. Will do!

File: test/integration/040_compare_test/test_compare.py [hunk 14]
```diff
@@ -2,13 +2,13 @@
...
 
-class TestConfigs(DBTIntegrationTest):
+class TestCompare(DBTIntegrationTest):
     @property
     def schema(self):
         return "config_040"
 
     def unique_schema(self):
-        return super(TestConfigs, self).unique_schema().upper()
+        return super(TestCompare, self).unique_schema().upper()
```
Comment (drewbanin) [inline test/integration/040_compare_test/test_compare.py]: blegh, just remove the `upper` here - my mistake!

Comment (mikekaminsky) [reply test/integration/040_compare_test/test_compare.py]: Ahhhhhhh

File: core/dbt/task/compare.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        problem_relation_list = []  # Get a list of relations to return
+
+        for relation_id in problems:
+            relation = database_relations_map[relation_id]
+            problem_relation_list.append(relation)
+            logger.info("{} {}".format(relation.type.upper(), relation))
+
+        return problem_relation_list
+
+        def interpret_results(self, results):
```
Comment (mikekaminsky) [inline core/dbt/task/compare.py]: @drewbanin is this indented in one level too far?

Comment (drewbanin) [reply core/dbt/task/compare.py]: @mikekaminsky yeah, good catch! You're going to want to back this up one level

**Comment by mikekaminsky** [comment]
Closing this (potentially temporarily) while thinking through the future of `dbt compare` and `dbt bootstrap` (https://github.com/fishtown-analytics/dbt/pull/1238)
